From a13970883e7b9cc7989efd7877534252c0317f06 Mon Sep 17 00:00:00 2001
From: Andre Vehreschild <vehre@gmx.de>
Date: Thu, 17 Apr 2025 11:57:47 +0200
Subject: [PATCH] Add more "beefy" tests.

Add tests that really make use of teams and not only a smoke test.
---
 CMakeLists.txt                           |   4 +-
 src/tests/unit/teams/CMakeLists.txt      |   2 +
 src/tests/unit/teams/teams_coindexed.f90 | 111 +++++++++++++++++++++++
 src/tests/unit/teams/test_teams_1.f90    |  66 ++++++++++++++
 4 files changed, 182 insertions(+), 1 deletion(-)
 create mode 100644 src/tests/unit/teams/teams_coindexed.f90
 create mode 100644 src/tests/unit/teams/test_teams_1.f90

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8f9ee35a..0dbb85a0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -818,9 +818,11 @@ if(opencoarrays_aware_compiler)
       add_caf_test(alloc_comp_multidim_shape 2 alloc_comp_multidim_shape)
       set_tests_properties(alloc_comp_multidim_shape PROPERTIES TIMEOUT 300)
     endif()
-    if(NOT CMAKE_Fortran_COMPILER_VERSION VERSION_LESS 15)
+    if(NOT CMAKE_Fortran_COMPILER_VERSION VERSION_LESS 16)
         add_caf_test(teams_this_image 8 teams_this_image)
         add_caf_test(teams_num_images 8 teams_num_images)
+        add_caf_test(test_teams_1 9 test_teams_1)
+        add_caf_test(teams_coindexed 2 teams_coindexed)
     endif()
   endif()
 
diff --git a/src/tests/unit/teams/CMakeLists.txt b/src/tests/unit/teams/CMakeLists.txt
index 8307b078..2ef2ee8e 100644
--- a/src/tests/unit/teams/CMakeLists.txt
+++ b/src/tests/unit/teams/CMakeLists.txt
@@ -10,4 +10,6 @@ caf_compile_executable(sync_team sync-team.f90)
 if (gfortran_compiler AND (NOT CMAKE_Fortran_COMPILER_VERSION VERSION_LESS 16.0.0))
     caf_compile_executable(teams_this_image teams_this_image.f90)
     caf_compile_executable(teams_num_images teams_num_images.f90)
+    caf_compile_executable(test_teams_1 test_teams_1.f90)
+    caf_compile_executable(teams_coindexed teams_coindexed.f90)
 endif()
diff --git a/src/tests/unit/teams/teams_coindexed.f90 b/src/tests/unit/teams/teams_coindexed.f90
new file mode 100644
index 00000000..5c3fab91
--- /dev/null
+++ b/src/tests/unit/teams/teams_coindexed.f90
@@ -0,0 +1,111 @@
+! BSD 3-Clause License
+!
+! Copyright (c) 2018-2025, Sourcery Institute
+! All rights reserved.
+!
+! Redistribution and use in source and binary forms, with or without
+! modification, are permitted provided that the following conditions are met:
+!
+! * Redistributions of source code must retain the above copyright notice, this
+!   list of conditions and the following disclaimer.
+!
+! * Redistributions in binary form must reproduce the above copyright notice,
+!   this list of conditions and the following disclaimer in the documentation
+!   and/or other materials provided with the distribution.
+!
+! * Neither the name of the copyright holder nor the names of its
+!   contributors may be used to endorse or promote products derived from
+!   this software without specific prior written permission.
+!
+! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+program teams_coindexed
+    use, intrinsic :: iso_fortran_env
+
+    type(team_type) :: parentteam, team, formed_team
+    integer :: t_num= 42, stat = 42, lhs
+    integer(kind=2) :: st_num=42
+    integer :: caf(2)[*]
+
+    parentteam = get_team()
+
+    caf = [23, 32]
+    form team(t_num, team)
+    form team(t_num, formed_team)
+
+    associate(me => this_image())
+        change team(team, cell[*] => caf(2))
+            ! for get_from_remote
+            ! Checking against caf_single is very limitted.
+            if (cell[me, team_number=t_num] /= 32) stop 1
+            if (cell[me, team_number=st_num] /= 32) stop 2
+            if (cell[me, team=parentteam] /= 32) stop 3
+
+            ! Check that team_number is validated
+            lhs = cell[me, team_number=5, stat=stat]
+            if (stat /= 1) stop 4
+
+            ! Check that only access to active teams is valid
+            stat = 42
+            lhs = cell[me, team=formed_team, stat=stat]
+            if (stat /= 1) stop 5
+
+            ! for send_to_remote
+            ! Checking against caf_single is very limitted.
+            cell[me, team_number=t_num] = 45
+            if (cell /= 45) stop 11
+            cell[me, team_number=st_num] = 46
+            if (cell /= 46) stop 12
+            cell[me, team=parentteam] = 47
+            if (cell /= 47) stop 13
+
+            ! Check that team_number is validated
+            stat = -1
+            cell[me, team_number=5, stat=stat] = 0
+            if (stat /= 1) stop 14
+
+            ! Check that only access to active teams is valid
+            stat = 42
+            cell[me, team=formed_team, stat=stat] = -1
+            if (stat /= 1) stop 15
+
+            ! for transfer_between_remotes
+            ! Checking against caf_single is very limitted.
+            cell[me, team_number=t_num] = caf(1)[me, team_number=-1]
+            if (cell /= 23) stop 21
+            cell[me, team_number=st_num] = caf(2)[me, team_number=-1]
+            ! cell is an alias for caf(2) and has been overwritten by caf(1)!
+            if (cell /= 23) stop 22
+            cell[me, team=parentteam] = caf(1)[me, team= team]
+            if (cell /= 23) stop 23
+
+            ! Check that team_number is validated
+            stat = -1
+            cell[me, team_number=5, stat=stat] = caf(1)[me, team_number= -1]
+            if (stat /= 1) stop 24
+            stat = -1
+            cell[me, team_number=t_num] = caf(1)[me, team_number= -2, stat=stat]
+            if (stat /= 1) stop 25
+
+            ! Check that only access to active teams is valid
+            stat = 42
+            cell[me, team=formed_team, stat=stat] = caf(1)[me]
+            if (stat /= 1) stop 26
+            stat = 42
+            cell[me] = caf(1)[me, team=formed_team, stat=stat]
+            if (stat /= 1) stop 27
+        end team
+
+        sync all
+        if (me == 1) print *, "Test passed."
+    end associate
+end program teams_coindexed
diff --git a/src/tests/unit/teams/test_teams_1.f90 b/src/tests/unit/teams/test_teams_1.f90
new file mode 100644
index 00000000..3d94c535
--- /dev/null
+++ b/src/tests/unit/teams/test_teams_1.f90
@@ -0,0 +1,66 @@
+! BSD 3-Clause License
+!
+! Copyright (c) 2018-2025, Sourcery Institute
+! All rights reserved.
+!
+! Redistribution and use in source and binary forms, with or without
+! modification, are permitted provided that the following conditions are met:
+!
+! * Redistributions of source code must retain the above copyright notice, this
+!   list of conditions and the following disclaimer.
+!
+! * Redistributions in binary form must reproduce the above copyright notice,
+!   this list of conditions and the following disclaimer in the documentation
+!   and/or other materials provided with the distribution.
+!
+! * Neither the name of the copyright holder nor the names of its
+!   contributors may be used to endorse or promote products derived from
+!   this software without specific prior written permission.
+!
+! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+program test_teams_1
+    use, intrinsic :: iso_fortran_env
+    use oc_assertions_interface, only : assert
+
+    integer :: caf(3,3)[*] != 42
+    type(team_type) :: row_team, column_team
+
+    caf = reshape((/(-i, i = 1, 9 )/), [3,3])
+    associate(me => this_image(), np => num_images())
+        call assert(np == 9, "I need exactly 9 teams.")
+
+        ! Form a row team
+        form team((me - 1) / 3 + 1, row_team, new_index=mod(me - 1, 3) + 1)
+        row_t: change team(row_team, row[*] => caf(:, team_number(row_team)))
+            ! Form column teams; each team has only one image
+            form team (team_number(), column_team)
+            col_t: change team(column_team, cell[*] => row(this_image()))
+                cell = team_number()
+                if (this_image() /= 1) row(this_image())[1] = cell
+            end team col_t
+            sync team(row_team)
+            if (this_image() == 1) caf(:, team_number(row_team))[1, team_number = -1] = row
+        end team row_t
+        sync all
+        if (me == 1) then
+            if (all(caf == reshape([1,1,1,2,2,2,3,3,3], [3, 3]))) then
+                print *, "Test passed."
+            else
+                print *, "Test failed."
+                print *, "Expected:", reshape([1,1,1,2,2,2,3,3,3], [3, 3])
+                print *, "Got     :", caf
+            end if
+        end if
+    end associate
+
+end program test_teams_1
